### PR TITLE
[FLINK-28322][table-api] DataStreamScan(Sink)Provider's new method is not compatible

### PR DESCRIPTION
**Pull request by JingsongLi**
## What is the purpose of the change

In [FLINK-25990](https://issues.apache.org/jira/browse/FLINK-25990) ,
Add a method "DataStream<RowData> produceDataStream(ProviderContext providerContext, StreamExecutionEnvironment execEnv)" in DataStreamScanProvider.
But this method has no default implementation, this is not compatible when users upgrade their DataStreamScanProvider implementation from 1.14 to 1.15. (The old method will not be called)

DataStreamSinkProvider is the same.

## Brief change log

DataStreamScanProvider's new method should be:
```
    default DataStream<RowData> produceDataStream(
            ProviderContext providerContext, StreamExecutionEnvironment execEnv) {
        return produceDataStream(execEnv);
    }
```

Impact on DataStreamSinkProvider:
DataStreamSinkProvider is no longer a functional interface and requires the use of an anonymous internal class.

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)

### CODE REVIEW

File: flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/connector/sink/DataStreamSinkProvider.java [hunk 1]
```diff
@@ -54,8 +54,10 @@
      *
      * @see SingleOutputStreamOperator#uid(String)
      */
-    DataStreamSink<?> consumeDataStream(
-            ProviderContext providerContext, DataStream<RowData> dataStream);
+    default DataStreamSink<?> consumeDataStream(
+            ProviderContext providerContext, DataStream<RowData> dataStream) {
+        return consumeDataStream(dataStream);
```
Comment (qingwei91) [inline flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/connector/sink/DataStreamSinkProvider.java:59]: Hi, can you elaborate what is the consequence of using this default implementation?

**Review by qingwei91 — APPROVED** [review]
Code looks good to me, I have a question around the implication of providing a default implementation.

PS: I am relatively new to the codebase, please let me know if my question does not makes sense

File: flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/connector/sink/DataStreamSinkProvider.java (hunk 1 above)
Comment (JingsongLi) [reply flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/connector/sink/DataStreamSinkProvider.java:59]: The old method will not be called.

**Review by twalthr — APPROVED** [review]
LGTM, I just had one question.

File: flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/connector/source/DataStreamScanProvider.java [hunk 2]
```diff
@@ -48,8 +48,10 @@ public interface DataStreamScanProvider extends ScanTableSource.ScanRuntimeProvi
      *
      * @see SingleOutputStreamOperator#uid(String)
      */
-    DataStream<RowData> produceDataStream(
-            ProviderContext providerContext, StreamExecutionEnvironment execEnv);
+    default DataStream<RowData> produceDataStream(
```
[the remark this answers is not in the document]
Comment (JingsongLi) [reply flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/connector/source/DataStreamScanProvider.java:51]: Yes, lambda will still fail, and downstream systems will still require code changes.
However, downstream systems can have the possibility to be compatible with 1.14 and 1.15, without having to maintain two copies of the code.

**Pull request merged**